app/questionnaire/routing/when_rule_evaluator.py

```diff
@@ -0,0 +1,63 @@
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Generator, Optional, Union
+
+from app.data_models import AnswerStore, ListStore
+from app.questionnaire import Location, QuestionnaireSchema
+from app.questionnaire.relationship_location import RelationshipLocation
+from app.questionnaire.routing.operator import OPERATIONS, Operator
+from app.questionnaire.value_source_resolver import (
+    ValueSourceResolver,
+    ValueSourceTypes,
+)
+
+
+@dataclass
+class WhenRuleEvaluator:
+    rule: dict
+    schema: QuestionnaireSchema
+    answer_store: AnswerStore
+    list_store: ListStore
+    metadata: dict
+    location: Union[Location, RelationshipLocation]
+    routing_path_block_ids: Optional[list] = None
+
+    def __post_init__(self) -> None:
+        self.value_source_resolver = ValueSourceResolver(
+            answer_store=self.answer_store,
+            list_store=self.list_store,
+            metadata=self.metadata,
+            schema=self.schema,
+            location=self.location,
+            list_item_id=self.location.list_item_id,
+            routing_path_block_ids=self.routing_path_block_ids,
+            use_default_answer=True,
+        )
+
+    def _evaluate(self, rule: dict[str, list]) -> Union[bool, Optional[datetime]]:
+        operator = Operator(next(iter(rule)))
+        operands = rule[operator.name]
+
+        if not isinstance(operands, list):
+            raise TypeError(
+                f"The rule is invalid. Got a non list type for operands - {type(operands)}"
```

Comment:
"The rule is invalid. Got a non list type for operands" -> "The rule is invalid, operands should be list type and not {type(operands)}"

Reply:
Updated with a slight change in [`ed7e6c4` (#629)](https://github.com/ONSdigital/eq-questionnaire-runner/pull/629/commits/ed7e6c44d22662e34cf62426400e5b7d9a315da1).